src/tensorlake/applications/function/application_call.py

```diff
@@ -1,11 +1,51 @@
-from typing import Any, List
+import inspect
+from typing import Any, Dict, List
 
-from ..interface import DeserializationError, File, Function
+from ..interface import DeserializationError, Function, SDKUsageError
 from ..metadata import ValueMetadata
-from .type_hints import function_arg_type_hint
+from .type_hints import coerce_to_type, function_arg_type_hint, function_signature
 from .user_data_serializer import deserialize_value, function_input_serializer
 
 
+def _get_application_param_count(application: Function) -> int:
+    """Returns the number of parameters for the application function, excluding 'self'."""
+    signature: inspect.Signature = function_signature(application)
+    params = list(signature.parameters.values())
+    # Exclude 'self' parameter for class methods
+    if len(params) > 0 and params[0].name == "self":
+        params = params[1:]
+    return len(params)
+
+
+def _coerce_payload_to_kwargs(
+    application: Function, payload: Dict[str, Any]
+) -> Dict[str, Any]:
+    """Coerces payload dict values to their expected types based on function signature.
+
+    Used for multi-parameter application functions.
+    """
+    signature: inspect.Signature = function_signature(application)
+    params = list(signature.parameters.values())
+
+    # Exclude 'self' parameter for class methods
+    if len(params) > 0 and params[0].name == "self":
+        params = params[1:]
+
+    kwargs: Dict[str, Any] = {}
+    for param in params:
+        if param.name in payload:
+            raw_value = payload[param.name]
+            kwargs[param.name] = coerce_to_type(raw_value, param.annotation)
+        elif param.default is not inspect.Parameter.empty:
+            kwargs[param.name] = param.default
+        else:
+            raise SDKUsageError(
```

Comment:
We can just call the function and see it fails. It's going to be a very regular Python error same as when user calls a function without passing all required parameters. I'm not happy about us duplicating regular Python logic in our code base.